finagle-stats-core/src/main/scala/com/twitter/finagle/stats/NonReentrantReadWriteLock.scala

```diff
@@ -23,10 +36,40 @@ private[stats] final class NonReentrantReadWriteLock extends AbstractQueuedSynch
   // we haven't provided an implementation for isHeldExclusively because the
   // docs for AQS advise implementors to provide implementations "as applicable"
   // and our use case doesn't require knowing when the lock is held exclusively.
+  private val SpinMax: Int = if (rwLockSpinWait() == -1) {
+    // Default number of spins for aarch64 machines set to 5 by default. 
+    // See below comment on reason.
+    if (System.getProperty("os.arch") == "aarch_64") {
+      5
+    } else {
+      1
+    }
+  } else {
+    rwLockSpinWait()
+  }
 
   override def tryAcquireShared(num: Int): Int = {
-    val cur = getState()
-    if (cur >= 0 && compareAndSetState(cur, cur + 1)) 1 else -1
+    var ret: Int = -1
+    var cur: Int = 0
+    var tries: Int = 0
+
+    // Try to acquire the lock up to SpinMax times before failing and 
+    // allowing AbstractQueuedSynchronizer to execute the slow-path lock.
+    // This is important on Arm systems where compareAndSetState may fail
+    // more often due to its weaker memory model.
+    do {
+      cur = getState()
+      if (cur >= 0 && compareAndSetState(cur, cur + 1)) {
+        ret = 1
+      } else {
+        // XXX: When JDK8 support is deprecated, onSpinWait() should be
```

Comment:
Mostly curious, does `XXX` have a special meaning here? If not, I would vote for removing it since we don't use this annotation elsewhere and it might confuse others when they read the comment.

Reply:
On another note, since this is on a hot path, wdyt optimizing the method as in the following, with `@tailrec`, also avoid usages of mutable vars:

```
override def tryAcquireShared(num: Int): Int = {
    // Try to acquire the lock up to SpinMax times before failing and
    // allowing AbstractQueuedSynchronizer to execute the slow-path lock.
    // This is important on Arm systems where compareAndSetState may fail
    // more often due to its weaker memory model.
    tryAcquireShared(getState(), -1, 1)
  }

  @tailrec
  private[this] def tryAcquireShared(cur: Int, ret: Int, tries: Int): Int = {
    if (cur >= 0 && compareAndSetState(cur, cur + 1)) 1
    else if (tries < SpinMax) tryAcquireShared(getState(), -1, tries + 1)
    else ret
  }
```

Reply:
To be honest, the entire purpose of this loop is not to be super-fast but consume some time to allow a previous lock holder to release without requiring an acquirer to synchronize via the kernel.  I'd like to to keep it as is if that is ok.

Reply:
I see, but we are iterating the path for up to 5 times, do you think that would help with release the pervious loc holder? With the recursive solution, it also helps shoving the allocations. This part of code will be called every time we add data to the stat, so it would be great to optimize it if possible.

Reply:
For the code in Stats, the critical sections are pretty short, and empirically on Arm, retrying allowed the system to successfully obtain the lock more often and not suffer the slow-path execution in AbstractQueuedSynchronizer.  

Testing with the micro-benchmarks and setting SpinMax to 1 had no impact on the runtime for x86.  

I'll test the suggested more elegant looking solution, though since I'm new to Scala the idea of trying to avoid explicit loops is a bit alien to me.

Reply:
Thanks for considering it, I did a bench marks for both execution time and allocations, with 1 iteration, there is no difference in execution time, but with 5 iterations, the non-recursive solution takes twice as much as time compared to the recursive solution.
But what worries me more is the allocation, with the non-recursive solution, every time user calls `add()` on a `Stat`, it will allocation 3 integers, which can be avoided with the recursive solution.

Reply:
@jyanJing , thanks for the explanation, little details like that help me understand better the concerns you had with the non-recursive solution.  I cleaned up your suggested fix and tested that it does no harm for my macro test case and the micro-benchmarks.  I think the PR is ready to be merged at this point.